### PR TITLE
Customize HTML/CSS template, support temporary comments

### PR DESCRIPTION
**Pull request by es94129**
Week 1

- customize template, use my own photos and introduction

Week 2
- display comments using GET
- users can POST comments

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -1,18 +1,141 @@
...
+                    </form> 
+                </p>
+            </section>
+        </div>
+        <div id="footer">
+            made on earth by a human <br> <a href="https://imfunniee.github.io">imfunniee</a>
+        </div>
+    <script src="script.js" type="text/javascript"></script>
+</body>
+</html>
```
Comment (ivancjw) [inline portfolio/src/main/webapp/index.html:141]: Pls leave a blank line at the end of the file.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -14,6 +14,9 @@
 
 package com.google.sps.servlets;
 
+import java.util.ArrayList;
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:17]: Pls sort the imports alphabetically.

File: portfolio/src/main/webapp/style.css [hunk 3]
```diff
@@ -1,9 +1,378 @@
...
+    background: #fff; 
+}
+::-webkit-scrollbar-thumb {
+    background: #0AB4B5; 
+}
+::-webkit-scrollbar-thumb:hover {
+    background: rgba(255, 50, 50,0.8); 
+}
+::selection {
+	color: #fff;
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:336]: use whitespace for indent.

File: portfolio/src/main/webapp/style.css [hunk 4]
```diff
@@ -1,9 +1,378 @@
...
+    padding:0px;
+    margin:0px;
+    background:linear-gradient(0deg,rgba(0,0,0,0.5),rgba(0,0,0,0.5)),url("https://images.unsplash.com/photo-1576422608509-27457a713964") center center;
+    background-size:cover;
+    width:100vw;
+    height:100vh;
+    overflow:hidden;
+    font-family: 'Josefin Sans', sans-serif;
+}
+#loading {
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:13]: always have a new line between two segments.

File: portfolio/src/main/webapp/index.html [hunk 5]
```diff
@@ -1,18 +1,141 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>Ying Chen</title>
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.min.css">
+    <link href="https://fonts.googleapis.com/css?family=Comfortaa:700" rel="stylesheet">
+    <script src="https://code.jquery.com/jquery-3.3.1.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
+    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.1/css/all.css" integrity="sha384-fnmOCqbTlWIlj8LyTjo7mOUStjsKC4pOpQbqyi7RrhN7udi9RwhKkMHpvLbHG9Sr" crossorigin="anonymous">
```
Comment (ivancjw) [inline portfolio/src/main/webapp/index.html:11]: Suggest to format it as the 

<script src="https://code.jquery.com/jquery-3.3.1.min.js" 
             integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" 
             crossorigin="anonymous">
</script>
 <link rel="stylesheet" 
          href="https://use.fontawesome.com/releases/v5.7.1/css/all.css" 
          integrity="sha384-fnmOCqbTlWIlj8LyTjo7mOUStjsKC4pOpQbqyi7RrhN7udi9RwhKkMHpvLbHG9Sr"
          crossorigin="anonymous">

The same below

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -23,10 +26,26 @@
...
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private ArrayList<String> messages = new ArrayList<String>();
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {    
+    String json = convertToJsonUsingGson(messages);
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: I think get might be called much more times than post. You might want to let the json (e.g. name it messages_json) to be a class member and set its value in doPost().

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -1,28 +1,70 @@
...
-      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!'];
-
-  // Pick a random greeting.
-  const greeting = greetings[Math.floor(Math.random() * greetings.length)];
-
-  // Add it to the page.
-  const greetingContainer = document.getElementById('greeting-container');
-  greetingContainer.innerText = greeting;
-}
+function showabout(){
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:1]: Naming convention: showAbout

The same below

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -1,28 +1,70 @@
...
+    $("#about_container").css("display","inherit");
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:2]: Whitespace between two elements

("display", "inherit")

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -1,28 +1,70 @@
...
+    $("#about_container").addClass("animated slideInLeft");
+    setTimeout(function(){
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:4]: setTimeout(function(){
            $("#about_container").removeClass("animated slideInLeft");
        },
        800);

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:1]: whitespace

function showabout() {

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -1,28 +1,70 @@
...
+      $("#box").css("display","none");
+      $("#about").removeClass("animated fadeIn");
+      $("#contact").removeClass("animated fadeIn");
+      $("#work").removeClass("animated fadeIn");
+    },1000);
+},1500);
+
+function getMessage() {
+    fetch('/data').then(response => response.json()).then((messages) => {
+        // document.getElementById('message_container').innerText = text;
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:57]: Remove

File: portfolio/src/main/webapp/style.css [hunk 11]
```diff
@@ -1,9 +1,378 @@
...
-
-#greeting-container {
-  margin-top: 20px;
-}
+@import url('https://fonts.googleapis.com/css?family=Josefin+Sans');
+
+body {
+    padding:0px;
+    margin:0px;
+    background:linear-gradient(0deg,rgba(0,0,0,0.5),rgba(0,0,0,0.5)),url("https://images.unsplash.com/photo-1576422608509-27457a713964") center center;
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:6]: whitespace

padding: 0px;

The same below

File: portfolio/src/main/webapp/style.css [hunk 12]
```diff
@@ -1,9 +1,378 @@
...
+    display:inline-block;
+}
+.box1 {
+    background:#0AB4B5;
+}
+.box2 {
+    background:#0AB4B5;
+    margin-left:-5px;
+}
+.circle_image{
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:56]: naming convention:

circle-image

File: portfolio/src/main/webapp/index.html [hunk 13]
```diff
@@ -1,18 +1,141 @@
...
+    <link rel="stylesheet" href="style.css">
+</head>
+<body>
+    <div id="loading">
+        <div id="spinner"></div>
+    </div>
+    <div id="particles-js"></div>
+    <div id="box">
+        <div class="box1 onlywide animated bounceOutLeft" style="animation-delay:1.7s;"></div>
```
Comment (ivancjw) [inline portfolio/src/main/webapp/index.html:20]: What does onlywide mean?

**Comment by es94129** [comment]
> Could you attach a screenshot to show you portfolio page?

Sure!
image
image
image
image

File: portfolio/src/main/webapp/index.html (hunk 13 above)
Comment (es94129) [reply portfolio/src/main/webapp/index.html]: Found that `onlywide` is useless, removed. Thanks for pointing out!

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -1,28 +1,82 @@
...
+
+function closeContact() {
+    $("#contact_container").addClass("animated slideOutDown");
+    setTimeout(function() {
+            $("#contact_container").removeClass("animated slideOutDown");
+            $("#contact_container").css("display", "none");
+        },
+        800);
+}
+setTimeout(function() {
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:54]: blank line above

Comment (ivancjw) [reply portfolio/src/main/webapp/script.js:54]: you may want to leverage some auto format tools, like

https://github.com/google/google-java-format

https://beautifier.io

File: portfolio/src/main/webapp/style.css [hunk 15]
```diff
@@ -1,9 +1,429 @@
...
+    cursor: pointer;
+    color: #0AB4B5;
+}
+
+.container {
+    width: 90vw;
+    height: 90vh;
+    padding: 5vh 5vw;
+    background: #0AB4B5;
+    /*background:#81D8D0;*/
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:209]: remove

File: portfolio/src/main/webapp/style.css [hunk 16]
```diff
@@ -1,9 +1,429 @@
...
+    color: #fff;
+    /*color:#0C1514;*/
```
Comment (ivancjw) [inline portfolio/src/main/webapp/style.css:211]: remove

File: portfolio/src/main/webapp/script.js [hunk 17]
```diff
@@ -1,28 +1,82 @@
...
+                createListElement(message));
+        });
+    });
+}
+
+function createListElement(text) {
+    const liElement = document.createElement('li');
+    liElement.innerText = text;
+    return liElement;
+}
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:82]: blank line at the end of file

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 18]
```diff
@@ -14,19 +14,44 @@
...
+    response.getWriter().println(messages_json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String message = request.getParameter("message");
+    messages.add(message);
+    messages_updated = true;
+    response.sendRedirect("/index.html");
+  }
```
[the remark this answers is not in the document]
Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: Yeah I also figured it out after this commit, thanks for the suggestion.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 19]
```diff
@@ -14,19 +14,44 @@
...
+        messages_updated = false;
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(messages_json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String message = request.getParameter("message");
```
[the remark this answers is not in the document]
Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: No, in `index.html`
```html
<textarea placeholder="your message" required rows="5" name="message">
```
the `required` keyword would force user to write at least 1 character in this field before hitting the "send" button. I tried it by myself.

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (es94129) [reply portfolio/src/main/webapp/script.js:54]: Actually I was using beautifier.io. After trying another time, the blank line is still not automatically added, so I decide to add it by myself. I think there might be some bug in this latter tool.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (ivancjw) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: The html and this java code are independent to each other. You should regard doGet and doPost as public APIs, and the html as the user of these APIs. In this sense, you should check user input in doPost. In team work, the java may be developed by you, and the html by another teammate.

One typical case that input check is not mandatory is local private functions. E.g. convertToJsonUsingGson. This function is only visible locally and called by local functions, you can think the input params are already validated.

FYI - You may want to use the response.sendError method

[the remark this answers is not in the document]
Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: I see. Now I use `response.sendError` and the page will show error message if the message is empty (tested by temporarily removing the `required` keyword).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 20]
```diff
@@ -38,9 +38,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String message = request.getParameter("message");
-    messages.add(message);
-    messages_json = convertToJsonUsingGson(messages);
-    response.sendRedirect("/index.html");
+    if (message.length() == 0) {
+        response.sendError(HttpServletResponse.SC_NOT_FOUND, "empty message");
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:42]: I think SC_FORBIDDEN makes more sense.

Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:42]: I didn't notice there is SC_FORBIDDEN, thanks for your suggestion!

**Pull request merged**